### ISSUE TITLE
Proposal: Use Pygments instead of highlight.js

### ISSUE THREAD

**Issue by mitya57** [mkdocs/mkdocs#1588]
I can see the following benefits of using Pygments:

- Does not require JavaScript, pure CSS (also means faster page loading).
- Works offline (the current release of mkdocs references a library from external website).
- Already implemented in Python-Markdown (`codehilite` extension).
- Easier to configure (e.g. `linenums`, `hl_lines` configure options).

I volunteer to make a pull request that:

- Automatically adds the Pygments CSS when the `codehilite` extension is in use (I will also try to take `pygments_style` option into account).
- Marks the `highlightjs`, `hljs_languages` and `hljs_style` options of the default themes as deprecated, with the suggestion to switch to `codehilite` extension.

@waylan Will you be open to such a pull request?

**Comment by waylan** [comment]
I would suggest you wrap that up in a plugin if you want everything automated. Although, personally I would just set `highlightjs: false` in my personal config, add the CSS file to `extra_css` and enable the `codehilite` extension. Its that easy, I don't see the need to wrap that up for people. Perhaps a [recipe][1] would suffice.

And users can get the CSS pre-made from [@richleland/pygments-css][2].

[1]: https://github.com/mkdocs/mkdocs/wiki/MkDocs-Recipes
[2]: https://github.com/richleland/pygments-css

**Comment by mitya57** [comment]
My idea was to make `codehilite` work automatically, without need for `extra_css`. This way we can offer Pygments as a first-class solution for code highlighting, and eventually deprecate `highlight.js` in favour of that, and clean up the code base (most of code for Pygments support will be in Python-Markdown, not in mkdocs).

It looks like you do not like my idea. I do not insist, but I would like to know your reasons — why prefer highlight.js over Pygments? In any case, if you think we should stick to `highlight.js`, I will be happy to provide a recipe in the wiki.

**Comment by waylan** [comment]
> I would like to know your reasons — why prefer highlight.js over Pygments?

Fair question. Here are my reasons:

1. This was not my choice but the choice of the project admins (of which I am not). I do not intend to override their choice regardless of any opinion I have. A search of old closed issues should turn up some old discussion(s) about this.

2. The `highlightjs`, `hljs_languages` and `hljs_style` options were just introduced in the 1.0 release.  It seems a little premature to be deprecating them already. In fact, we have promised to support the existing API for the 1.x series. Any deprecation would be a 2.0 feature. Given we just released 1.0, I'm not ready to start having that conversation. 

3. The `highlightjs`, `hljs_languages` and `hljs_style` options were introduced specifically so a user could easily disable highlight.js to use pygments or some other preferred choice. The source files are being pulled from a CDN so the files would not continue to remain part of the built site. In other words, I already included the feature to the extent I believe it should be included. A "recipe" might be the only improvement to the existing situation.

4. Notice that the options are not MkDocs options, but theme options, which means no code execution happens, only template logic in the theme. In other words the MkDocs code base knows nothing about these settings. It is all template controlled. That being the case, were would your code go to prevent the pygments styles from being included if highlight.js is disabled? That would require a plugin.

As an aside, the mkdocs-nature theme (maintained by me) uses pygments (with the Markdown extension) and does not include the highlight.js options at all. That is certainly a reasonable choice for a theme. It is simply not the choice made by the MkDocs team for the built-in themes.
